### DISCUSSION TITLE
Adversarial training mnist

### DISCUSSION THREAD

**Discussion by thungp** [Trusted-AI/adversarial-robustness-toolbox#1288] [Q&A] [answered]
Ref:
(a)
https://github.com/Trusted-AI/adversarial-robustness-toolbox/blob/main/notebooks/adversarial_training_mnist.ipynb

(b) https://github.com/Trusted-AI/adversarial-robustness-toolbox/issues/367

Could someone tell me, if there is Adversarial training example for this notebook that uses TF 2.x libraries?

Additionally, has anyone by chance adopted this sample code to work
with Resnet50 and/or inceptionv3 model?

I made an attempt to upgrade the code to use TF 2.x  packages.
Summary of some of he key changes attempted

https://colab.research.google.com/drive/1vxdj8kYrZI9d6SDdfubGA4m5U_EgqdbR?usp=sharing
But getting this error.
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-19-70de71d5aabd> in <module>
      7 trainer = AdversarialTrainer(robust_classifier, attacks, ratio=1.0)
      8 # dir(trainer)
----> 9 trainer.fit(x_train, y_train, nb_epochs=83, batch_size=50)

~/.conda/envs/art/lib/python3.6/site-packages/art/defences/trainer/adversarial_trainer.py in fit(self, x, y, batch_size, nb_epochs, **kwargs)
    243 
    244                 # Fit batch
--> 245                 self._classifier.fit(x_batch, y_batch, nb_epochs=1, batch_size=x_batch.shape[0], verbose=0, **kwargs)
    246                 attack_id = (attack_id + 1) % len(self.attacks)
    247 

~/.conda/envs/art/lib/python3.6/site-packages/art/estimators/classification/classifier.py in replacement_function(self, *args, **kwargs)
     69                 if len(args) > 0:
     70                     args = tuple(lst)
---> 71                 return fdict[func_name](self, *args, **kwargs)
     72 
     73             replacement_function.__doc__ = fdict[func_name].__doc__

~/.conda/envs/art/lib/python3.6/site-packages/art/estimators/classification/tensorflow.py in fit(self, x, y, batch_size, nb_epochs, **kwargs)
    946         if self._train_step is None:
    947             raise TypeError(
--> 948                 "The training function `train_step` is required for fitting a model but it has not been " "defined."
    949             )
    950 

TypeError: The training function `train_step` is required for fitting a model but it has not been defined.
```

I'm only speculating at this point, before I potentially dive deeper, 
could it be that the AdversarialTrainer class needs to get upgraded to TF2.x? (maybe leveraging https://www.tensorflow.org/guide/upgrade)?
or hopefully maybe some error on my part?

-Peter

**Comment by beat-buesser** [comment, marked as answer]
Hi @thungp To train a model with `TensorFlowV2Classifier` you need to define a function defining the training step for the model and provide it to argument `train_step` of `TensorFlowV2Classifier` similar as in this example script: https://github.com/Trusted-AI/adversarial-robustness-toolbox/blob/main/examples/get_started_tensorflow_v2.py

**Reply by thungp** [reply]
Thank you!

**Reply by thungp** [reply]
Question:
Is the code that was used to generate the mnist model with 1024 vs 128 units in the 1st dense layer, saved off somewhere?  Is that shareable?

path = get_file('mnist_cnn_robust.h5', extract=False, path=config.ART_DATA_PATH,
                url='https://www.dropbox.com/s/yutsncaniiy5uy8/mnist_cnn_robust.h5?dl=1')

I am assuming just that Dense layer with units 1024 vice 128 was swapped in with all other layer weighs untouched?  Was the model then just retrained with just that layer marked trainable?
How many epochs? Just one?  What was the optimizer learning rates? 
After retraining the model was the overall model fine tuned with all the layers made trainable for x amount of epochs?  What was the criteria that was used to prevent overfitting?  

When you pass in the updated model (with 1024), was that model passed into the
AdversarialTrainer and then retrained again, with all layers marked trainable? 

Are there some intermediary statistics while adversarially training the robust_classifier to help determine the ideal number of epochs to run the adversarial trainer?  

Would like to be able to recreate the robust classifier to be able to recreate the performance graph  in[16].

In one attempt I made, this was the result:

![image](https://user-images.githubusercontent.com/235944/131964011-059a1daa-202d-4100-b897-2f0f295552aa.png)

But I feel, I may not have either recreated the model, correctly and/or retrained the new layer properly, perhaps.

**Reply by beat-buesser** [reply]
Hi @thungp Thank you! I think your figure looks correct and as I would expect it. The figure in cell 16 of the notebook (https://github.com/Trusted-AI/adversarial-robustness-toolbox/blob/main/notebooks/adversarial_training_mnist.ipynb) needs to be updated because it looks like the robustness is achieved by gradient masking caused by an overconfident classifier. We will update it with issue #1298.

**Reply by thungp** [reply]
Hi @beat-buesser ,

Ref:
(a) https://arxiv.org/abs/1706.06083?source=post_page---------------------------


1. Are you at liberty to share the code that shows how this saved model
`path = get_file('mnist_cnn_robust.h5', extract=False, path=config.ART_DATA_PATH,
                url='https://www.dropbox.com/s/yutsncaniiy5uy8/mnist_cnn_robust.h5?dl=1')`

referenced from the current 1.7.1 baseline (https://github.com/Trusted-AI/adversarial-robustness-toolbox/blob/main/notebooks/adversarial_training_mnist.ipynb):
that has the 1024 dense layer was created from the original mnist model?

I am assuming that will confirm if the changed 1024 layer and the subsequent top dense layer with units 10, were new layers with randomized weights and presumably needed to be retrained, with the prior layer weights preserved (using transfer learning technique vice, retraining the whole model from scratch?) (see https://keras.io/guides/transfer_learning/)
but since retraining was using original mnist training set (and not a new data set)  presumably no fine tuning of the model was done (which would have made all layers trainable flag to false and do a final retraining (fitting) to eek out any additional incremental improvements of the model, but some how be cautious about over fitting.

2. Another question that I had were the comments in the code:
# We had performed this before, starting with a randomly intialized model.
Did you basically fully retrain the mnist model from scratch vice apply some transfer learning technique which freezes and preserves  the [:-2] layer weights?


3. When you say that the figure in cell 16 needs to be updated as it appears like robustness was achieved by gradient masking caused by an overconfident classifier, is that based on this paper, https://arxiv.org/pdf/1706.06083.pdf
pg 14 figure 6(a) coupled with 
this line of code: 
attacks = BasicIterativeMethod(robust_classifier, eps=0.3, eps_step=0.01, max_iter=40)
that shows the eps=0.3 and that is roughly where the performance starts to drop
and in the sentence that states, "We observe that for smaller # than the one used during training the models achieve equal or
higher accuracy, as expected. For MNIST, we notice a large drop in robustness for slightly large epsilon (eps in code) values, potentially due to the fact that the threshold operators learned are tuned to the exact value of epsilon (eps in code) used during training (Appendix C)." [a]


4. Also, i am presuming that by upgrading code to TF 2.3 (swapping out the wrappers with KerasClassifier and TensorFlowV2Classifier shoudl yield similar, results, correct?  With your presumed update, if you have the time to confirm it does work, would appreciate it.   Based on your referenced example script.
5. https://github.com/Trusted-AI/adversarial-robustness-toolbox/blob/main/examples/get_started_tensorflow_v2.py
I made an initial attempt at replicating the results locally 

used the example  train_step as reference:
`# Per https://github.com/Trusted-AI/adversarial-robustness-toolbox/discussions/1288
# need to define the training step (train_step) and provide it into the TensorflowV2Classifier

optimizer = tf.keras.optimizers.Adam(learning_rate=0.01)

def train_step(model, images, labels):
    with tf.GradientTape() as tape:
        predictions = model(images, training=True)
        loss = loss_object(labels, predictions)
    gradients = tape.gradient(loss, model.trainable_variables)
    optimizer.apply_gradients(zip(gradients, model.trainable_variables))
    
loss_object = tf.losses.SparseCategoricalCrossentropy(from_logits=False)

robust_classifier = TensorFlowV2Classifier(input_shape=input_shape, model=robust_classifier_model,
                                           nb_classes=num_classes, clip_values=(min_, max_),loss_object=loss_object,
                                           train_step=train_step)` 

Set up the attack as before:
attacks = BasicIterativeMethod(robust_classifier, eps=0.3, eps_step=0.01, max_iter=40)

# Evaluate the robust classifier's performance on the original test data:
`x_test_robust_pred = np.argmax(robust_classifier.predict(x_test[:100]), axis=1)
nb_correct_robust_pred = np.sum(x_test_robust_pred == np.argmax(y_test[:100], axis=1))

print("Original test data (first 100 images):")
print("Correctly classified: {}".format(nb_correct_robust_pred))
print("Incorrectly classified: {}".format(100-nb_correct_robust_pred))`

Results:
`Original test data (first 100 images):
Correctly classified: 11
Incorrectly classified: 89`

Which appears to me, that I something happened to the model in adversarially training it where the model lost it's classification effectiveness on non adversarially images.

And because of that, naturally the following results make sense:

# Evaluate the robust classifier's performance on the adversarial test data (white-box setting):
`attacker_robust = FastGradientMethod(robust_classifier, eps=0.5)
x_test_adv_robust = attacker_robust.generate(x_test[:100])

x_test_adv_robust_pred = np.argmax(robust_classifier.predict(x_test_adv_robust), axis=1)
nb_correct_adv_robust_pred = np.sum(x_test_adv_robust_pred == np.argmax(y_test[:100], axis=1))

print("Adversarial test data (first 100 images):")
print("Correctly classified: {}".format(nb_correct_adv_robust_pred))
print("Incorrectly classified: {}".format(100-nb_correct_adv_robust_pred))

# Results: 
Adversarial test data (first 100 images):
Correctly classified: 7
Incorrectly classified: 93
`
And subsequent code and plots:

`
eps_range = [0.01, 0.02, 0.03, 0.04, 0.05, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9]
nb_correct_original = []
nb_correct_robust = []

for eps in eps_range:
    attacker.set_params(**{'eps': eps})
    attacker_robust.set_params(**{'eps': eps})
    x_test_adv = attacker.generate(x_test[:100])
    x_test_adv_robust = attacker_robust.generate(x_test[:100])
    
    x_test_adv_pred = np.argmax(classifier.predict(x_test_adv), axis=1)
    nb_correct_original += [np.sum(x_test_adv_pred == np.argmax(y_test[:100], axis=1))]
    
    x_test_adv_robust_pred = np.argmax(robust_classifier.predict(x_test_adv_robust), axis=1)
    nb_correct_robust += [np.sum(x_test_adv_robust_pred == np.argmax(y_test[:100], axis=1))]

eps_range = [0] + eps_range
nb_correct_original = [nb_correct_pred] + nb_correct_original
nb_correct_robust = [nb_correct_robust_pred] + nb_correct_robust

fig, ax = plt.subplots()
ax.plot(np.array(eps_range), np.array(nb_correct_original), 'b--', label='Original classifier')
ax.plot(np.array(eps_range), np.array(nb_correct_robust), 'r--', label='Robust classifier')

legend = ax.legend(loc='upper center', shadow=True, fontsize='large', bbox_to_anchor=(0.5, -0.15), ncol= 2)
legend.get_frame().set_facecolor('#00FFCC')

plt.xlabel('Attack strength (eps)')
plt.ylabel('Correct predictions')
plt.show()

`

![image](https://user-images.githubusercontent.com/235944/132049500-f8825094-d377-4976-bd0a-c015580e9ecd.png)

PS. note just slight change to the location of the legend so it did not obscure performance curves.

**Reply by thungp** [reply]
As an update, I believe my TF 2.x (actually 2.3) test was using too high a learning rate .01. I changed it back to the default .001 when retraining the modified model (with dense layer 1024).
I retested the classification  of the model prior to adversarial training of the model and I am now getting this result after testing against the original mnist test dataset.
Original test data (first 100 images):
Correctly classified: 100
Incorrectly classified: 0

Expecting better results now after adversarially training that model and would expect it to be inline with the TF 1.x recreation of the example tutorial. 

Unfortunately, for reasons unknown at this time, although using a Titan V GPU,
it is taking about 20 minutes per iteration (epoch) for adversrarial training.
I saw in the git comments, adversarial training took about 80 minutes 
on an NVIDIA 100. 

I tried to do  look up some performance comparisons
and based on this link.
https://www.gpuzoo.com/Compare/NVIDIA_Tesla_V100_SMX2__vs__NVIDIA_Titan_V/
it didn't seem too different.
I am curious if you guys by chance are scaling the training across multiple GPUs
to get it down to 80 minutes.

I may need to look into that. https://towardsdatascience.com/how-to-scale-training-on-multiple-gpus-dae1041f49d2
For now though, I'll just be patient and let it take the ~ 27 hours to finish.

Update. I realized I did not set the modified mnist model (retrained model layers prior the dense layer to trainable = True).
I believe (please feel free to confirm), that when adversarially training a model, all the trainable layers should be set to trainable
except any BatchNormailzation layers (as per https://keras.io/guides/transfer_learning/), but with mnist, there isn't any). 


Additional update:
After modifying the model dense layer 128->1024, 
I decided to retest the model both against, normal images and adversarially attacked images just to make sure 
things looked similar prior to modifying the dense layer.

Interestingly with eps = 0.5
I was still getting this:
Adversarial test data (first 100 images) (TF 2.3 TensorFlowV2Classifier) :
Correctly classified: 83
Incorrectly classified: 17
vice
Adversarial test data (first 100 images) (TF 1.15 Keras Classifier) :
Correctly classified: 22
Incorrectly classified: 78

I did a visual on the adversarial samples generated:
![image](https://user-images.githubusercontent.com/235944/132110676-b5b6ca90-0957-4e92-aa3c-e501347003bb.png)


I recall this thread reporting an issue on vanishing gradient, which I am suspecting may be why it appears that not all the images are being attacked.

I see this thread discussing it. 
https://github.com/Trusted-AI/adversarial-robustness-toolbox/issues/367

And I see you ran the model predicting logits instead of probabilities.

@beat-buesser  I dn't suppose you still have the example code please feel free to share. 
also, I'll be looking it up and trying that out.

**Reply by thungp** [reply]
@beat-buesser 
I modified the script to define the classifier like this:
`optimizer = tf.keras.optimizers.Adam(learning_rate=0.001)

def train_step(model, images, labels):
    with tf.GradientTape() as tape:
        predictions = model(images, training=True)
        loss = loss_object(labels, predictions)
    gradients = tape.gradient(loss, model.trainable_variables)
    optimizer.apply_gradients(zip(gradients, model.trainable_variables))
// end of function defition, note the python spacing were not preserved in this post    

loss_object = tf.keras.losses.CategoricalCrossentropy(from_logits=True)

modified_classifier = TensorFlowV2Classifier(input_shape=input_shape, model=modified_classifier_model,
                                           nb_classes=num_classes, clip_values=(min_, max_),loss_object=loss_object,
                                           train_step=train_step)`

Regenerated some adversarial samples;
`eps_value = 0.5
modified_attacker  = FastGradientMethod(modified_classifier, eps=eps_value) # `
![image](https://user-images.githubusercontent.com/235944/132367970-5c595c73-f25d-42ac-a2ca-8014894635e4.png)

And it looks more like what I expected.

Now in the process of adversarially retraining the model anfter ensurng all layers were marked trainable (please confirm if that assumption is correct, or not")

Taking about 908.32s/it ( or about  ~ 15 per iteration).  Expecting similar response curves to the TF 1.15 recreation experiment. 
Hopefully should be done later today.

**Reply by beat-buesser** [reply]
Hi @thungp Thank you very much for sharing your work! Let me try to answer the questions first:

2. I couldn't find any scripts anymore that trained the model used in this notebook. It looks like the original contributor only shared the notebook and the model.
3. Yes, the figure in the notebook should look more like Figure 6 of https://arxiv.org/pdf/1706.06083.pdf. I think it is masking gradients because it looks to have a very high or even perfect confidence which results in a categorical entropy loss that is numerically zero and therefore also its loss gradients are numerically zero and that's why the attack doesn't work, but other attack approaches would work on this model.
4. Yes, both frameworks should achieve the same results.

- Yes, usually all layers should be trainable for adversarial training if starting training from scratch. ART calls the respective frameworks by setting the most likely defaults for trainability depending on the tasks and depending on if the framework requires it or not using the optional argument `training_mode` in the supporting estimator's methods. For special task-specific models like ASR ART's estimators contain special internal methods to set batch-norm layers, etc. correctly.

 The image above looks good, there is no more gradient masking as all images have perturbations. What are the benign and adversarial accuracies on the training and test sets?
 
 Would you be interested to share your results to update the notebook and fix #1298?

**Reply by beat-buesser** [reply]
@thungp 

About the issue of slow iterations, what is your batch-size during training and attack?

About the gradient masking in the first image of MNIST images, you could calculate the cross entropy loss for images that show perturbations and for images that don't, you should see that the latter is numerically very low.

**Reply by thungp** [reply]
@beat-buesser ,

Will get back to you on your other question and responses, but, just wanted to post the result of the TF 2.3
results across the range of EPS values:
![image](https://user-images.githubusercontent.com/235944/132398581-b0af1dce-76cc-4f72-a166-9f6b5a6f8387.png)

The originally classifier results seem a little too good and is tracking the robust classifier curve a little too closely IMHO.

I feel, like I may need to go over my script a little closely.

**Reply by thungp** [reply]
> @thungp
> 
> About the issue of slow iterations, what is your batch-size during training and attack?
> 
> About the gradient masking in the first image of MNIST images, you could calculate the cross entropy loss for images that show perturbations and for images that don't, you should see that the latter is numerically very low.

Batch_size = 50
trainer.fit(x_train, y_train, nb_epochs=83, batch_size=50) # baseline

**Reply by beat-buesser** [reply]
@thungp I agree with your opinion on the last post figure. Would you be able to share an update version of your notebook to try it out?

One thing we noticed previously is that adversarial training depends strongly on the hyper-parameters. I would recommend to start with exactly the same parameter values as in  https://arxiv.org/pdf/1706.06083.pdf.

**Comment by thungp** [comment]
@beat-buesser ,
https://colab.research.google.com/drive/1TqhB_Gsc3Q23rh1wHmlRL296__l9BeKn?usp=sharing

1. I looked at the  article, and tried to match the pareameter values. Not sure if you have time to double check.
2. In the original example art,  there was a comment about the following:
"# We had performed this before, starting with a randomly intialized model.
# Adversarial training takes about 80 minutes on an NVIDIA V100.
# The resulting model is the one loaded from mnist_cnn_robust.h5 above."

I guess I must have miissed that, because as you can see in my example, I assumed that I would be starting out
with a pretrained model and when changing the configuration of one of the dense layers, I would want to try and 
treat it like transer learning. and would want to have some weights to the model before attemppting to train it further
with with adversarial samples.

If the whole model is supposed to be randomly1 initialized,  it then begs the question, how is it randomly initialized?
Also, ifyou have any exmaples of how to save the model that is wrapped after adversarially trained,
please feel free to send some sample code. 
robust_classifier = modified_classifier 

Also training is takng almost 22 hours on Google Colab, vice the 80 minutes that the example is saying which is also making
me question stuff.
![image](https://user-images.githubusercontent.com/235944/135920264-c818e5cd-5a1a-438b-bebb-51530c09bb61.png)

**Reply by beat-buesser** [reply]
Hi @thungp Thank you for sharing your results. I have tried to open the Colab notebook, but I had to request access first, could you please grant access?

Is `eps` in infinity norm?

I think the red line looks quite similar as the blue line in Figure 6a of Madry et al. (https://arxiv.org/pdf/1706.06083.pdf).

The blue line in  your plot should converge to zero, and decrease much earlier. Can you check if gradient masking occurs?

About the training time, what is your batch size? How busy is the GPU during the 22 hours?

Most users access to he estimator property `model` (e.g. `robust_classifier.model`) and use the tools of the corresponding ML framework to save the model.

**Reply by thungp** [reply]
@beat-buesser 
I think I fixed the link to allow anyone with the link to be able to view the notebook.
https://colab.research.google.com/drive/1TqhB_Gsc3Q23rh1wHmlRL296__l9BeKn?usp=sharing

Else, I think I would need to get your email address to to grant you specific access.
When you get the chance, please try the link [https://colab.research.google.com/drive/1TqhB_Gsc3Q23rh1wHmlRL296__l9BeKn?usp=sharing] again.

**Reply by beat-buesser** [reply]
@thungp Thank you! As you have mentioned above, to get started with adversarial training it would definitely be easier if you don't have the frozen layers. I would recommend to exactly reproduce the experiments of Madry et al. (https://arxiv.org/pdf/1706.06083.pdf, their experiment code is available on GitHub too). For these experiments, I also would not work with `mnist_cnn_original.h5` at all and start to implement the model completely from scratch instead, it will not require much code and be easier to initialise, etc.
